src/cryptonote_core/service_node_list.cpp

```diff
@@ -1553,59 +1554,73 @@ namespace service_nodes
+    auto const serialize_version                  = data_for_serialization::get_version(hf_version);
+
+    for (data_for_serialization *serialize_entry : data)
+    {
+      if (serialize_entry->version != serialize_version) m_long_term_states_added_to = true;
+      serialize_entry->version = serialize_version;
+      serialize_entry->clear();
+    }
+
     {
       std::lock_guard<boost::recursive_mutex> lock(m_sn_mutex);
-      data.quorum_states.reserve(m_old_quorum_states.size());
+      short_term_data.quorum_states.reserve(m_old_quorum_states.size());
       for (const quorums_by_height &entry : m_old_quorum_states)
-        data.quorum_states.push_back(serialize_quorum_state(hf_version, entry.height, entry.quorums));
+        short_term_data.quorum_states.push_back(serialize_quorum_state(hf_version, entry.height, entry.quorums));
 
-      if (m_state_history.size() == 1)
-      {
-        data.states.push_back(serialize_service_node_state_object(hf_version, *m_state_history.begin()));
-      }
-      else if (m_state_history.size() >= 2)
+      uint64_t const min_short_term_height = m_state.height - MAX_SHORT_TERM_STATE_HISTORY;
+      uint64_t const max_short_term_height = m_state.height - MIN_DERIVABLE_SHORT_TERM_STATE_OFFSET;
+      if (m_long_term_states_added_to)
       {
-        size_t num_to_reserve = 1; // recent state
-        if (m_state_history.size() > MAX_SHORT_TERM_STATE_HISTORY)
-          // + 10k historic blocks (-1 because the oldest short-term history might also be historic)
-          num_to_reserve += m_state_history.size() - (MAX_SHORT_TERM_STATE_HISTORY - 1);
-        data.states.reserve(num_to_reserve);
-
-        for (auto it = m_state_history.cbegin(), nextit = std::next(it); nextit != m_state_history.cend(); it = nextit++)
+        for (auto it : m_state_history)
         {
-          state_t const &curr = *it, &next = *nextit;
-
-          data.states.push_back(serialize_service_node_state_object(hf_version, curr));
-          uint64_t height_delta = next.height - curr.height;
-          if (height_delta != STORE_LONG_TERM_STATE_INTERVAL)
-          {
-            // TODO(loki): Preserve the quorum state from v403 until they all get purged from storage then we can remove that check.
-            if (next.is_migrated_from_v403() && next.height != m_state_history.rbegin()->height)
-              continue;
-
-            data.states.push_back(serialize_service_node_state_object(hf_version, next));
-            break;
-          }
+          if (it.height >= min_short_term_height) break;
+          long_term_data.states.push_back(serialize_service_node_state_object(hf_version, it));
         }
       }
+
+      for (auto it = m_state_history.lower_bound(min_short_term_height);
+           it != m_state_history.end();
+           it++)
+      {
+        if (it->height > max_short_term_height) break;
+        short_term_data.states.push_back(serialize_service_node_state_object(hf_version, *it));
+      }
```

Comment:
Are you sure? Lower bound gives iterator to the first element >= min_short_term_height or end() if it doesn't exist.

Here I want all the states inbetween the range, and actually just their quorums except the last one. So that find approach would gives jut one?

Reply:
Yes, you're right!  I didn't notice the difference between `min` in the lower_bound and `max` in the test, my bad.